pypiserver/backend.py

```diff
@@ -0,0 +1,224 @@
+        "relfn_unix",  # The relative file path in unix notation
+        "parsed_version",  # The package version as a tuple of parts
+        "digester",  # a function that calculates the digest for the package
+    ]
+
+    def __init__(
+        self, pkgname, version, fn=None, root=None, relfn=None, replaces=None
+    ):
+        self.pkgname = pkgname
+        self.pkgname_norm = normalize_pkgname(pkgname)
+        self.version = version
+        self.parsed_version = parse_version(version)
+        self.fn = fn
+        self.root = root
+        self.relfn = relfn
+        self.relfn_unix = None if relfn is None else relfn.replace("\\", "/")
+        self.replaces = replaces
+        self.digest = None
+
+    def __repr__(self):
+        return "{}({})".format(
+            self.__class__.__name__,
+            ", ".join(
+                [
+                    f"{k}={getattr(self, k, 'AttributeError')!r}"
+                    for k in sorted(self.__slots__)
+                ]
+            ),
+        )
+
+    @property
+    def fname_and_hash(self):
+        if self.digest is None:
+            self.digest = self.digester(self)
+        hashpart = f"#{self.digest}" if self.digest else ""
+        return self.relfn_unix + hashpart
+
+
+class Backend:
+    def __init__(self, config: Configuration):
+        self.hash_algo = config.hash_algo
+
+    def get_all_packages(self) -> t.Iterable[PkgFile]:
+        """Implement this method to return an Iterable of all packages (as
+        PkgFile objects) that are available in the Backend.
+        """
+        raise NotImplementedError
+
+    def add_package(self, filename: str, fh: t.BinaryIO):
+        """Add a package to the Backend. `filename` is the package's filename
+        (without any directory parts). It is just a name, there is no file by
+        that name (yet). `fh` is an open file object that can be used to read
+        the file's content. To convert the package into an actual file on disk,
+        run `as_file(filename, fh)`.
+        """
+        raise NotImplementedError
+
+    def remove_package(self, pkg: PkgFile):
+        """Remove a package from the Backend"""
+        raise NotImplementedError
+
+    def exists(self, filename) -> bool:
+        """Does a package by the given name exist?"""
+        raise NotImplementedError
+
+    def digest(self, pkg: PkgFile):
+        if self.hash_algo is None:
+            return None
+        return digest_file(pkg.fn, self.hash_algo)
+
+    def get_projects(self) -> t.Iterable[str]:
+        """Return an iterable of all (unique) projects available in the store
+        in their PEP503 normalized form. When implementing a Backend class,
+        either use this method as is, or override it with a more performant
+        version.
+        """
+        normalized_pkgnames = set()
+        for x in self.get_all_packages():
+            if x.pkgname:
+                normalized_pkgnames.add(x.pkgname_norm)
```

Comment:
Is this check for the `pkgname` necessary? It seems like, since it' a required positional argument to the `PkgFile`, we should be able to rely on it (and the normalized name) always being present, in which case we could do something like

```py
yield from set(pkg.pkgname_norm for pkg in self.get_all_packages())
```

Reply:
Yeah this extra check is quite redundant